crypto/src/tests/secp256k1_tests.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2021, Facebook, Inc. and its affiliates
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::*;
+use crate::{
+    secp256k1::{Secp256k1KeyPair, Secp256k1PrivateKey, Secp256k1PublicKey, Secp256k1Signature},
+    traits::EncodeDecodeBase64,
+};
+
+use rand::{rngs::StdRng, SeedableRng as _};
+use signature::{Signature, Signer, Verifier};
+
+pub fn keys() -> Vec<Secp256k1KeyPair> {
+    let mut rng = StdRng::from_seed([0; 32]);
+
+    (0..4).map(|_| Secp256k1KeyPair::generate(&mut rng)).collect()
+}
+
+#[test]
+fn serialize_deserialize() {
+    let kpref = keys().pop().unwrap();
+    let public_key = kpref.public();
+
+    let bytes = bincode::serialize(&public_key).unwrap();
+    let pk2 = bincode::deserialize::<Secp256k1PublicKey>(&bytes).unwrap();
+    assert_eq!(*public_key, pk2);
+
+    let private_key = kpref.private();
+    let bytes = bincode::serialize(&private_key).unwrap();
+    let privkey = bincode::deserialize::<Secp256k1PrivateKey>(&bytes).unwrap();
+    let bytes2 = bincode::serialize(&privkey).unwrap();
+    assert_eq!(bytes, bytes2);
+}
+
+#[test]
+fn import_export_public_key() {
+    let kpref = keys().pop().unwrap();
+    let public_key = kpref.public();
+    let export = public_key.encode_base64();
+    let import = Secp256k1PublicKey::decode_base64(&export);
+    assert!(import.is_ok());
+    assert_eq!(&import.unwrap(), public_key);
+}
+
+#[test]
+fn import_export_secret_key() {
+    let kpref = keys().pop().unwrap();
+    let secret_key = kpref.private();
+    let export = secret_key.encode_base64();
+    let import = Secp256k1PrivateKey::decode_base64(&export);
+    assert!(import.is_ok());
+    assert_eq!(import.unwrap().as_ref(), secret_key.as_ref());
+}
+#[test]
+fn to_from_bytes_signature() {
+    let kpref = keys().pop().unwrap();
+    let signature = kpref.sign(b"Hello, world");
+    let sig_bytes = signature.as_ref();
+    let rebuilt_sig = Secp256k1Signature::from_bytes(sig_bytes).unwrap();
+    assert_eq!(rebuilt_sig.sig, signature.sig);
+}
+
+#[test]
+fn verify_valid_signature() {
+    // Get a keypair.
+    let kp = keys().pop().unwrap();
+
+    // Make signature.
+    let message: &[u8] = b"Hello, world!";
+    let digest = message.digest();
+
+    let signature = kp.sign(&digest.0);
+
+    // Verify the signature.
+    assert!(kp.public().verify(&digest.0, &signature).is_ok());
+}
+
+#[test]
+fn verify_invalid_signature() {
+    // Get a keypair.
+    let kp = keys().pop().unwrap();
+
+    // Make signature.
+    let message: &[u8] = b"Hello, world!";
+    let digest = message.digest();
+
+    let signature = kp.sign(&digest.0);
+
+    // Verify the signature.
+    let bad_message: &[u8] = b"Bad message!";
+    let digest = bad_message.digest();
+
+    assert!(kp.public().verify(&digest.0, &signature).is_err());
+}
+
```

Comment:
It would be great to test the main cryptographic functions using [secp256k1](https://docs.rs/secp256k1/latest/secp256k1/), the rust bindings to the Bitcoin lib. One way to engineer those tests is with [proptest](https://docs.rs/proptest/latest/proptest/index.html) which will generate random inputs, let you define how to generate a random key with each lib, then test the comparisons.
The nice thing about this being that secp256k1 only appears in dev-dependencies of this package, and that it's probably not hard to compare that implementation to secp256k1 in micro-benchmarks.

Reply:
almost had it working, but due to the different representations of a signature in the c library, comparing the actual signature result was a bit challenging to massage them into. 

will follow up with the prop test PR